### PR TITLE
UI Changes For Attributes

### PR DESCRIPTION
**Pull request by Vijay-Giri**

### CODE REVIEW

**Comment by Vijay-Giri** [comment]
CSS Pending.

File: templatestore/frontend/src/pages/templateScreen.js [hunk 1]
```diff
@@ -57,6 +57,10 @@ class TemplateScreen extends Component {
         this.saveTemplate = this.saveTemplate.bind(this);
         this.setMandatoryAttributes = this.setMandatoryAttributes.bind(this);
         this.onVersionAliasChange = this.onVersionAliasChange.bind(this);
+        this.getAttributes = this.getAttributes.bind(this);
+        this.buildOptions = this.buildOptions.bind(this);
```
Comment (wilspi) [inline templatestore/frontend/src/pages/templateScreen.js:61]: what is `builldOptions`

File: templatestore/views.py [hunk 2]
```diff
@@ -250,6 +232,53 @@ def post_template_view(request):
...
+                )
+
+                invalid_valued_attributes = invalid_valued_attributes | set(
+                    key
+                    for key in ts_settings.TE_TEMPLATE_ATTRIBUTES.keys()
+                    if key in data["attributes"]
+                    and "allowed_values" in ts_settings.TE_TEMPLATE_ATTRIBUTES[key]
+                    and data["attributes"][key]
+                    not in ts_settings.TE_TEMPLATE_ATTRIBUTES[key]["allowed_values"]
+                )
```
Comment (wilspi) [inline templatestore/views.py:271]: what is being done here?

File: templatestore/views.py [hunk 3]
```diff
@@ -617,13 +646,21 @@ def patch_attributes_view(request, name):
...
+            cfgs = TemplateConfig.objects.filter(type=template.type)
+
             missing_attributes = set(
-                ts_settings.TE_TEMPLATE_ATTRIBUTES_KEYS
+                ts_settings.TE_TEMPLATE_ATTRIBUTES.keys()
+            ).difference(set(data["attributes"].keys()))
+
+            missing_attributes = missing_attributes | set(
+                cfgs[0].attributes.keys()
             ).difference(set(data["attributes"].keys()))
```
Comment (wilspi) [inline templatestore/views.py:662]: what?

File: templatestore/models.py [hunk 4]
```diff
@@ -40,7 +40,9 @@ def clean(self):
                 }
             )
         # add default attributes
-        for k in ts_settings.TE_TEMPLATE_ATTRIBUTES_KEYS:
+        for k in [
+            attributes for attributes in ts_settings.TE_TEMPLATE_ATTRIBUTES.keys()
+        ]:
```
Comment (wilspi) [inline templatestore/models.py:45]: `for k in ts_settings.TE_TEMPLATE_ATTRIBUTES.keys()`

File: templatestore/frontend/src/pages/templateScreen.js (hunk 1 above)
Comment (Vijay-Giri) [reply templatestore/frontend/src/pages/templateScreen.js:61]: buildOptions is a function which takes a list of values that are allowed for a particular attribute and returns options for the select element.

File: templatestore/views.py (hunk 2 above)
Comment (Vijay-Giri) [reply templatestore/views.py:271]: Here we check that all the attributes (that has a fixed set of values to choose from) have valid values. At first we check that if "allowed_values" are there in the config attributes then is the value given by user from the given set of values or not. Same thing is done for general mandatory attributes (like lob and journey). So invalid_valued_attributes contains a set of attributes that have invalid values. If the set in not empty then we raise an exception that these attributes have invalid values.

File: templatestore/views.py (hunk 3 above)
Comment (Vijay-Giri) [reply templatestore/views.py:662]: this checks that no mandatory attribute is missing in the attributes given by the user.

**Pull request merged**